src/libcore/slice.rs

```diff
@@ -295,22 +295,23 @@ impl<T> SliceExt for [T] {
     fn binary_search_by<F>(&self, mut f: F) -> Result<usize, usize> where
         F: FnMut(&T) -> Ordering
     {
-        let mut base : usize = 0;
-        let mut lim : usize = self.len();
+        let mut base = 0usize;
+        let mut s = self;
 
-        while lim != 0 {
-            let ix = base + (lim >> 1);
-            match f(&self[ix]) {
-                Equal => return Ok(ix),
+        loop {
+            let (head, tail) = s.split_at(s.len() >> 1);
+            if tail.is_empty() {
+                return Err(base)
+            }
+            match f(&tail[0]) {
```

Comment:
I do not know why the code should not be the same.
Looking at the IR, the LLVM basic block that generates that jump seems to be
```llvm
"_ZN5slice12_$u5b$T$u5d$8split_at21h15129678983694567549E.exit": ; preds = %loop_body
  %5 = getelementptr inbounds i32, i32* %s.sroa.0.0.ph, i32 %3
  %6 = icmp eq i32 %s.sroa.7.0, %3
  %switchtmp = icmp eq i32* %5, null
  %or.cond = or i1 %6, %switchtmp
  br i1 %or.cond, label %match_case, label %match_case4
```